File: C/11052.c
#include <stdio.h>  // 카드 구매하기

int max(int a, int b)
{
	if (a > b)
		return a;
	else
		return b;
}

int dp[10000] = { 0, };

int main()
{
	int n;
	scanf("%d", &n);

	int arr[10000] = { 0, };
	for (int i = 1; i <= n; i++)
		scanf("%d", &arr[i]);

	for (int i = 1; i <= n; i++)
	{
		for (int j = 1; j <= i; j++)
		{
			dp[i] = max(dp[i], dp[i - j] + arr[j]);
		}
	}
	printf("%d", dp[n]);

	return 0;
}

File: JAVA/src/_11052.java
import java.io.*;

// 카드 구매하기
public class _11052 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int n = Integer.parseInt(br.readLine());
        int[] arr = new int[n + 1];
        String[] input = br.readLine().split(" ");
        for (int i = 1; i <= n; i++) {
            arr[i] = Integer.parseInt(input[i - 1]);
        }

        int[] ans = new int[n + 1];
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= i; j++) {
                ans[i] = Math.max(ans[i], ans[i - j] + arr[j]);
            }
        }
        bw.write(ans[n] + "");

        bw.flush();
        bw.close();
    }
}
